lib/src/gauge_group.rs

```diff
@@ -3,46 +3,76 @@ use std::{
     sync::{Arc, Weak},
 };
 
+use crate::pipeline::aggregation::Aggregation;
+use crate::pipeline::aggregator::DimensionedMeasurementsMap;
 use crate::{
     gauge::{self, StatisticSetGauge},
-    pipeline::aggregation::statistic_set::StatisticSet,
+    pipeline::aggregator::DimensionPosition,
     types::Name,
 };
 
 #[derive(Default)]
 pub struct GaugeGroup {
-    gauges: HashMap<Name, Weak<StatisticSetGauge>>,
+    pub(crate) dimensioned_gauges:
+        HashMap<DimensionPosition, HashMap<Name, Weak<StatisticSetGauge>>>,
 }
 
 impl GaugeGroup {
     /// Create or retrieve a gauge.
     pub fn gauge(&mut self, name: impl Into<Name>) -> Arc<StatisticSetGauge> {
+        self.dimensioned_gauge(name, Default::default())
+    }
+
+    pub fn dimensioned_gauge(
+        &mut self,
+        name: impl Into<Name>,
+        dimensions: DimensionPosition,
+    ) -> Arc<StatisticSetGauge> {
         let name = name.into();
-        match self.gauges.get(&name) {
+        let gauge_position = self.dimensioned_gauges.entry(dimensions).or_default();
+        match gauge_position.get(&name) {
             Some(existing) => match existing.upgrade() {
                 Some(existing) => existing,
                 None => {
                     let gauge: Arc<StatisticSetGauge> = Arc::new(gauge::statistic_set_gauge());
-                    self.gauges.insert(name, Arc::downgrade(&gauge));
+                    gauge_position.insert(name, Arc::downgrade(&gauge));
                     gauge
                 }
             },
             None => {
                 let gauge: Arc<StatisticSetGauge> = Arc::new(gauge::statistic_set_gauge());
-                self.gauges.insert(name, Arc::downgrade(&gauge));
+                gauge_position.insert(name, Arc::downgrade(&gauge));
                 gauge
             }
         }
     }
 
-    pub fn reset(&mut self) -> impl Iterator<Item = (Name, StatisticSet)> + '_ {
-        self.gauges.retain(|_name, gauge| gauge.upgrade().is_some());
+    // reset gauge group, aggregating and returning previous values
+    pub fn reset(&mut self) -> DimensionedMeasurementsMap {
+        self.dimensioned_gauges
+            .retain(|_dimension_position, gauges| {
+                gauges.retain(|_name, gauge| gauge.upgrade().is_some());
+                !gauges.is_empty()
+            });
 
-        self.gauges.iter().filter_map(|(name, possible_gauge)| {
-            possible_gauge
-                .upgrade()
-                .and_then(|gauge| gauge.reset())
-                .map(|statistic_set| (name.to_owned(), statistic_set))
-        })
+        self.dimensioned_gauges
+            .iter()
+            .map(|(dimension_position, gauges)| {
```

Comment:
how is this removing dimension positions that are no longer populated with any gauges once you filter_map them out?

Reply:
good call, fixed in [c6a833a](https://github.com/kvc0/goodmetrics_rs/pull/13/commits/c6a833a801d30f08f0bf673ac3e1dcf4d1478825) and also added a test